src/shared_realm.cpp

```diff
@@ -554,5 +555,107 @@ util::Optional<int> Realm::file_format_upgraded_from_version() const
     return util::none;
 }
 
+#define VERSION_CAST(version_id) {(version_id).version, (version_id).index}
+
+// Precondition: `m_version` is not greater than `new_version`
+// Postcondition: `m_version` is equal to `new_version`
+void Realm::HandoverPackage::advance_to_version(VersionID new_version)
+{
+    if ((SharedGroup::VersionID)VERSION_CAST(new_version) == (SharedGroup::VersionID)VERSION_CAST(m_version_id)) {
+        return;
+    }
+    REALM_ASSERT_DEBUG(((SharedGroup::VersionID)VERSION_CAST(new_version) >
+                        (SharedGroup::VersionID)VERSION_CAST(m_version_id)));
+
+    // Open `Realm` at handover version
+    _impl::RealmCoordinator& coordinator = get_coordinator();
+    Realm::Config config = coordinator.get_config();
+    config.cache = false;
+    SharedRealm realm = coordinator.get_realm(config);
+    REALM_ASSERT(!realm->is_in_read_transaction());
+    realm->m_group = &const_cast<Group&>(realm->m_shared_group->begin_read(VERSION_CAST(m_version_id)));
+
+    // Import handover, advance version, and then repackage for handover
+    auto objects = realm->accept_handover(std::move(*this));
+    transaction::advance(*realm->m_shared_group, realm->m_binding_context.get(),
+                         realm->m_config.schema_mode, VERSION_CAST(new_version));
+    *this = realm->package_for_handover(objects);
+}
+
+Realm::HandoverPackage::~HandoverPackage()
+{
+    if (is_awaiting_import()) {
+        get_coordinator().get_realm()->m_shared_group->unpin_version(VERSION_CAST(m_version_id));
+        mark_not_awaiting_import();
+    }
+}
+
+Realm::HandoverPackage Realm::package_for_handover(std::vector<AnyThreadConfined> objects_to_hand_over)
+{
+    verify_thread();
+    if (is_in_transaction()) {
+        throw InvalidTransactionException("Cannot package handover during a write transaction.");
+    }
+
+    HandoverPackage handover;
+    auto version_id = m_shared_group->pin_version();
+    handover.m_version_id = VERSION_CAST(version_id);
+    handover.m_source_realm = shared_from_this();
+    // Since `m_coordinator` is used to determine if we need to unpin when destroyed,
+    // `m_coordinator` should only be set after `pin_version` succeeds in case it throws.
+
+    handover.m_objects.reserve(objects_to_hand_over.size());
+    for (auto &object : objects_to_hand_over) {
+        REALM_ASSERT(object.get_realm().get() == this);
+        handover.m_objects.push_back(object.export_for_handover());
+    }
+
+    return handover;
+}
+
+std::vector<AnyThreadConfined> Realm::accept_handover(Realm::HandoverPackage handover)
+{
+    REALM_ASSERT(handover.is_awaiting_import()); // Enforced by move semantics
+    auto unpin_version = util::make_scope_exit([&]() noexcept {
+        m_shared_group->unpin_version(VERSION_CAST(handover.m_version_id));
+        handover.mark_not_awaiting_import();
+    });
+
+    verify_thread();
```

Comment:
Thread check needs to be before unpin_version is set up, as unpinning the version isn't thread-safe.